### PR TITLE
Replace non-ellipsis default arguments

### PR DESCRIPTION
**Pull request by srittau**
`=None` was replaced with `: Optional[Any] = ...`, other defaults with their respective types.

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
I'm not sure I agree with this. Having `x=None` in the stub signals to future editors of the file that nobody really thought about the type, so they should investigate what the correct type is. But `x: Optional[Any] = ...` suggests that somebody already did the work and decided that `Any` is the right type, so our hypothetical future editor would not investigate the type further, even though the correct type is probably narrower than `Any`.

Similarly, the correct type for parameters with a non-None default isn't necessarily the same as the type of the default. For example, maybe an `x=0` parameter actually allows `x` to be of type `float`.

**Comment by srittau** [comment]
I was under the impression that mypy would infer a function's type from the default value if no type annotation is given, but that does not seem to be the case. So, just replacing the default by `...` would not make the type situation worse.

On the other hand, basically all the stubs that used non-ellipsis defaults were very rudimentary typed, and probably mostly automatically generated. They often have no type information or already use `Any` as placeholder. I don't think future editors would assume the `Any`s to be the correct type already, and using `Optional` here is at least a little step in the right direction.

I will go over the not-`None` defaults again, but since I manually changed most of them, I am fairly confident that the vast majority of them is correct. It was mostly `charset=''`, `index=0`, and `some_flag=True` arguments.

**Comment by srittau** [comment]
I have now reviewed all defaults that were not marked with `None` and fixed them. Most important fix was marking `charset` and `errors` with `Text` instead of `str` in werkzeug.

File: third_party/2/itsdangerous.pyi [hunk 1]
```diff
@@ -138,13 +138,13 @@ class TimedJSONWebSignatureSerializer(JSONWebSignatureSerializer):
     DEFAULT_EXPIRES_IN = ...  # type: int
     expires_in = ...  # type: int
     def __init__(self, secret_key: bytes_like, expires_in: Optional[int] = ..., **kwargs) -> None: ...
-    def make_header(self, header_fields=Optional[MutableMapping]) -> MutableMapping: ...
+    def make_header(self, header_fields: Optional[MutableMapping]) -> MutableMapping: ...
```
Comment (JelleZijlstra) [inline third_party/2/itsdangerous.pyi]: Nice catch.

File: third_party/2and3/werkzeug/contrib/cache.pyi [hunk 2]
```diff
@@ -1,83 +1,84 @@
...
-    def set_many(self, mapping, timeout=None): ...
+    def set(self, key, value, timeout: Optional[float] = ...): ...
+    def add(self, key, value, timeout: Optional[float] = ...): ...
+    def set_many(self, mapping, timeout: Optional[float] = ...): ...
     def delete_many(self, *keys): ...
     def has(self, key): ...
     def clear(self): ...
-    def inc(self, key, delta=1): ...
-    def dec(self, key, delta=1): ...
+    def inc(self, key, delta=...): ...
```
Comment (JelleZijlstra) [inline third_party/2and3/werkzeug/contrib/cache.pyi]: Why didn't you add types here?

**Comment by JelleZijlstra** [comment]
I just went over the whole patch and it struck me that it's also just a really big change, and many of the changes are at least a little risky. (For example, any method change can affect base classes, and it's possible you got some of the argument types wrong.) I don't like having so many substantial changes in a single PR—if something does go wrong, it makes it harder to figure out where the problem is coming from. I don't want to give you more work though by splitting up this PR, so maybe let's keep this open for a bit to see what other typeshed maintainers think.

**Comment by srittau** [comment]
It would probably make sense to split the `=None` changes from the other changes, especially since the former make up the bulk of the patch and the latter are now hand-checked. Also, a few libraries (werkzeug, tk) make up the bulk of the patch. It would be easy to move them to separate PRs.

**Comment by srittau** [comment]
Or of course replace `=None` with just `= ...` for now. While I prefer `= Optional[Any]` to make future patches a tad easier, I do not feel strongly about that.

**Comment by ilevkivskyi** [comment]
FWIW I just checked this PR against our internal codebases and it is clean.

**Pull request merged**